### ISSUE TITLE
Misaligned output on aarch64-darwin

### ISSUE THREAD

**Issue by ners** [maralorn/nix-output-monitor#78]
I'm not quite sure why, but the overview section of the `nom build` output is always misaligned.

<img width="640" alt="image" src="https://user-images.githubusercontent.com/50560955/209471160-e101a1c9-a139-4841-8690-ee4997a5269b.png">

This was tested on an M1 MacBook Pro running macOS Ventura.

**Comment by ners** [comment]
Also spotted in x86-64 WSL.

![grafik](https://user-images.githubusercontent.com/50560955/210280700-d08f54d5-4c7f-446d-9cef-3d5761f50e79.png)

**Comment by maralorn** [comment]
To be fair, while on my system it’s not that bad as above I am also having misaligned output issues.  This is super hard to get right on all platforms simultaneously, the problem is that it’s really hard to predict how many cells a certain emoji is wide in the current font.
In foot, my current terminal of choice, this actually changes during a run of nom (i.e. during a waiting build the hourglass sometimes has width 1 and sometimes width 2 without a discernible pattern …)

**Comment by 999eagle** [comment]
This sounds like it's caused in kitty by https://github.com/kovidgoyal/kitty/issues/3998. Kitty has a different behavior for character widths, so the sequence `U+23F3 U+FE0E` (hourglass + VS15) is rendered with a width of one cell instead of the two cells NOM calculates.
This rendering difference means it's generally impossible to do alignment the "correct" way across all terminal emulators without special casing.

**Comment by maralorn** [comment]
Yes, that is also my fear. But even reliably detecting in which terminal we run seems hard.

I am considering to actually drop all emojis which tend to have an irregular width or at least offer an option to do so …

**Comment by maralorn** [comment]
Interesting. I observed the "stuff changes if there is a whitespace behind it" behavior as well. 
I think I am consistently using a presentation selector.

This discussion gives me hope, that there is a principled solution to this. Will investigate when I have the time.

**Comment by takeda** [comment]
Not sure if this provides any insight, but I just noticed this.

I'm on mac and use iTerm. Whenever I use NOM the output is always misaligned (it looks like the picture in initial comment). I'm currently connected to NixOS from the same iTerm on the same Mac and the output looks properly aligned (nom runs inside `screen`, not sure if that matters, did not had opportunity to run it without)

So it likely might have to do with how the terminal is initialized, perhaps the selector (mentioned by @zeorin) is set differently on those systems.

**Comment by amjoseph-nixpkgs** [comment]
I observe this as well under `alacritty` and the `PragmataPro` font (with nerdfonts glyphs).

This tool is so awesome!  I can't believe it took me this long to try it.

**Comment by 9999years** [comment]
I've noticed that the `wcwidth` call here frequently returns `-1`:

https://github.com/maralorn/nix-output-monitor/blob/23d4302bab647e9f0f238ff3566f33b13ed9209e/lib/NOM/Print/Table.hs#L75

The main culprits are these characters:

- `⏱`
- `⏳`
- `⏵`

However, patching `widthFold` to return a minimum of 1 doesn't help much.

Another problem is that the output includes variation selectors 15 and 16 to force text and emoji presentation (respectively) of the subsequent codepoint; for some reason, `wcwidth` thinks these characters are 1 cell wide.

**Comment by maralorn** [comment]
@amjoseph-nixpkgs Do you mind sharing on which system you encountered this issue? I am getting the impression that wcwidth works correctly on linux but is broken when used on Mac and WSL.

**Comment by amjoseph-nixpkgs** [comment]
> @amjoseph-nixpkgs Do you mind sharing on which system you encountered this issue?

Linux, `sway`, Nixpkgs

I don't run NixOS (no systemd) but I don't think that has much to do with this issue.
